py/gc.c

```diff
@@ -477,29 +477,21 @@ static void gc_deal_with_stack_overflow(void) {
     }
 }
 
-static void gc_sweep(void) {
-    #if MICROPY_PY_GC_COLLECT_RETVAL
-    MP_STATE_MEM(gc_collected) = 0;
-    #endif
-    // free unmarked heads and their tails
-    int free_tail = 0;
-    #if MICROPY_GC_SPLIT_HEAP_AUTO
-    mp_state_mem_area_t *prev_area = NULL;
-    #endif
+// Run finalisers for all to-be-freed blocks first
+// (necessary during gc_sweep_all() to avoid accessing already-freed memory)
+static void sweep_run_finalisers(void) {
+    #if MICROPY_ENABLE_FINALISER
     for (mp_state_mem_area_t *area = &MP_STATE_MEM(area); area != NULL; area = NEXT_AREA(area)) {
-        size_t end_block = area->gc_alloc_table_byte_len * BLOCKS_PER_ATB;
-        if (area->gc_last_used_block < end_block) {
-            end_block = area->gc_last_used_block + 1;
-        }
-
-        size_t last_used_block = 0;
-
-        for (size_t block = 0; block < end_block; block++) {
-            MICROPY_GC_HOOK_LOOP(block);
-            switch (ATB_GET_KIND(area, block)) {
-                case AT_HEAD:
-                    #if MICROPY_ENABLE_FINALISER
-                    if (FTB_GET(area, block)) {
+        assert(area->gc_last_used_block <= area->gc_alloc_table_byte_len * BLOCKS_PER_ATB);
+        // Small speed optimisation: skip over empty FTB blocks
+        size_t ftb_end = (area->gc_last_used_block + BLOCKS_PER_FTB) / BLOCKS_PER_FTB;
+        for (size_t ftb_idx = 0; ftb_idx < ftb_end; ftb_idx++) {
```

Comment:
This arithmetic is a little tricky, considering `gc_last_used_block` is inclusive.

I think `(area->gc_last_used_block + BLOCKS_PER_FTB) / BLOCKS_PER_FTB` is equivalent to `area->gc_last_used_block / BLOCKS_PER_FTB + 1`, so the above two lines are equivalent to:
```c
        size_t ftb_end = area->gc_last_used_block / BLOCKS_PER_FTB;
        for (size_t ftb_idx = 0; ftb_idx <= ftb_end; ftb_idx++) {
```
That might be a little clearer when reading the code (although the compiler is probably already making this optimisation).

Reply:
Have changed! Oddly, this did change something in the code generation on -Os (got slightly bigger) :shrug:

Reply:
Looking at the latest edits on commit https://github.com/micropython/micropython/pull/16015#issuecomment-2412939085, it seems that 3 bare-metal ports got smaller by 8 bytes, which is good, and I assume is due to this change.